### PR TITLE
Migrate storage APIs to Bytes32

### PR DESCRIPTION
**Pull request by shemnon**
## PR description

The current internal storage APIs implement UInt256 rather than Bytes32, which is an accident of prior EVM design.  This migrates all internal handling of storage addresses and values to Bytes32.

The main performance gain is that UInt256 stores the data as an array of ints, whereas Bytes32 keeps a sliced or wrapped array of bytes.  Since database APIs are all byte array based this should remove unneeded conversions when going to/from the DB.

Signed-off-by: Danno Ferrin <danno.ferrin@swirldslabs.com>

## Documentation

- [x] I thought about documentation and added the `doc-change-required` label to this PR if
    [updates are required](https://wiki.hyperledger.org/display/BESU/Documentation).

## Changelog

- [x] I thought about the changelog and included a [changelog update if required](https://wiki.hyperledger.org/display/BESU/Changelog).

### CODE REVIEW

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/bonsai/BonsaiWorldStateUpdater.java [hunk 1]
```diff
@@ -201,7 +182,7 @@ public void commit() {
                 (keyHash, entryValue) -> {
                   final Hash slotHash = Hash.wrap(keyHash);
                   if (!deletedStorageUpdates.containsKey(slotHash)) {
-                    final UInt256 value = UInt256.fromBytes(RLP.decodeOne(entryValue));
+                    final Bytes32 value = Bytes32.leftPad(Bytes.wrap(RLP.decodeOne(entryValue)));
```
Comment (pinges) [inline ethereum/core/src/main/java/org/hyperledger/besu/ethereum/bonsai/BonsaiWorldStateUpdater.java:185]: Why do we need the Bytes.wrap here?

File: ethereum/core/src/test/java/org/hyperledger/besu/ethereum/privacy/PrivateStateGenesisAllocatorTest.java [hunk 2]
```diff
@@ -146,8 +146,8 @@ private void assertGenesisAccountApplied() {
   private void assertManagementContractApplied() {
     Account managementProxy = worldState.get(FLEXIBLE_PRIVACY_PROXY);
     assertThat(managementProxy.getCode()).isEqualTo(PROXY_RUNTIME_BYTECODE);
-    assertThat(managementProxy.getStorageValue(UInt256.ZERO))
-        .isEqualTo(UInt256.fromBytes(DEFAULT_FLEXIBLE_PRIVACY_MANAGEMENT));
+    assertThat(managementProxy.getStorageValue(Bytes32.ZERO))
+        .isEqualTo(Bytes32.leftPad(Bytes.wrap(DEFAULT_FLEXIBLE_PRIVACY_MANAGEMENT)));
```
Comment (pinges) [inline ethereum/core/src/test/java/org/hyperledger/besu/ethereum/privacy/PrivateStateGenesisAllocatorTest.java:150]: DEFAULT_FLEXIBLE_PRIVACY_MANAGEMENT is of type Address, which implements Bytes. Why do we need the wrap?

**Review by pinges — COMMENTED** [review]
Please have a look at my comments.

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/bonsai/BonsaiWorldStateUpdater.java (hunk 1 above)
Comment (shemnon) [reply ethereum/core/src/main/java/org/hyperledger/besu/ethereum/bonsai/BonsaiWorldStateUpdater.java]: We don't.  Large scale change missed it.

File: ethereum/core/src/test/java/org/hyperledger/besu/ethereum/privacy/PrivateStateGenesisAllocatorTest.java (hunk 2 above)
Comment (shemnon) [reply ethereum/core/src/test/java/org/hyperledger/besu/ethereum/privacy/PrivateStateGenesisAllocatorTest.java]: We don't, removing it.

**Comment by shemnon** [comment]
I looked for all instances where `Bytes.wrap(Bytes ... bytes)` was called with a single argument and removed them all.

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/bonsai/TrieLogLayer.java [hunk 3]
```diff
@@ -136,13 +136,13 @@ public static TrieLogLayer readFrom(final RLPInput input) {
...
-        final Map<Hash, BonsaiValue<UInt256>> storageChanges = new TreeMap<>();
+        final Map<Hash, BonsaiValue<Bytes32>> storageChanges = new TreeMap<>();
         input.enterList();
         while (!input.isEndOfCurrentList()) {
           input.enterList();
           final Hash slotHash = Hash.wrap(input.readBytes32());
-          final UInt256 oldValue = nullOrValue(input, RLPInput::readUInt256Scalar);
-          final UInt256 newValue = nullOrValue(input, RLPInput::readUInt256Scalar);
+          final Bytes32 oldValue = nullOrValue(input, RLPInput::readUInt256Scalar);
+          final Bytes32 newValue = nullOrValue(input, RLPInput::readUInt256Scalar);
```
Comment (fab-10) [inline ethereum/core/src/main/java/org/hyperledger/besu/ethereum/bonsai/TrieLogLayer.java:145]: is reading uint256 correct here, or should it read bytes32?

**Review by fab-10 — COMMENTED** [review]
LGTM, just checking if a reading operation is correct.
Have you any result on how much this can improve performance?

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/bonsai/TrieLogLayer.java (hunk 3 above)
Comment (shemnon) [reply ethereum/core/src/main/java/org/hyperledger/besu/ethereum/bonsai/TrieLogLayer.java]: `readBytes32Scalar` wasn't in the RLPInput interface, so I pulled it up.  It also exposed that the proofs were still on `UInt256`s

**Comment by shemnon** [comment]
Benchmarks I have are not for public consumption.

This was mostly about pulling the `UInt256` out of APIs it's not needed (i.e. no math calls).  Each UInt256 requires a round trip from a primitive byte array into a primitive int array and then back to a bytes array for serialization.  So this is more about GC pressure (where the byte array can be re-used and passed around as slices, instead of two primitive conversions minimum) that it was about wall clock time.

**Review by fab-10 — APPROVED** [review]

**Pull request merged**